sql: implement function OID reference

chengxiong-ruan opened the pull request:
This commit implements `[FUNCTION xxxx]` OID references syntax of functions. With this change, functions can be called with the new OID numerical representation. For example `SELECT [FUNCTION 123]('helloworld')`. The intention of this syntax is only for internal serialization of references to UDFs. But it's general enough for builtin functions as well.

A new implementation of the `ResolvableFunctionReference` interface, `OIDFunctionReference` is introduced for function resolution purpose for the new syntax
The `ResolveFunctionByOID` method is also refactored to return a qualified name.

Fixes: #83231

Release note: None

chengxiong-ruan committed:
sql: implement function OID reference

This commit implements `[FUNCTION xxxx]` OID references syntax
of functions. With this change, functions can be called with
the new OID numerical representation. For example
`SELECT [FUNCTION 123]('helloworld')`. The intention of this
syntax is only for internal serialization of references to UDFs.
But it's general enough for builtin functions as well.

A new implementation of the `ResolvableFunctionReference`
interface, `OIDFunctionReference` is introduced for function
resolution purpose for the new syntax.

The `ResolveFunctionByOID` method is also refactored to return
a qualified name.

Fixes: #83231

Release note: None

DrewKimball reviewed (commented):
> Reviewed 26 of 27 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96442)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @mgartner, and @shermanCRL)

___
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NNc-MxN4aSKDS8itC2G:-NNc-MxN4aSKDS8itC2H:b-9g6d7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8525d699015f276792ee3068373ae9d86d91a37/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L41)):*
> ```
> 
> query I
> SELECT [FUNCTION $fn_oid]()
> ```

Can you add a case where we attempt to use this syntax from a different database? And maybe also one or two where the function has been altered or dropped.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NNc-MxN4aSKDS8itC2G:-NNcBZax0ZjhvGlMgdDE:brc3u0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8525d699015f276792ee3068373ae9d86d91a37/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L41)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Can you add a case where we attempt to use this syntax from a different database? And maybe also one or two where the function has been altered or dropped.
</blockquote></details>

Good call! Yeah, we need to disallow cross-db references event referencing by id. Fixed that and added test case for it.
Also add cases for renaming a function, altering function's parent schema and dropping a function


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
I'd defer to @mgartner for approval, but <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-:-NNcP1QvAn0z7kjqMEPc:bxyy4tb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @mgartner, and @shermanCRL)

___
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NNc-MxN4aSKDS8itC2G:-NNcOUmx7SV6PRRFeAHN:b-atx7l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8525d699015f276792ee3068373ae9d86d91a37/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L41)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Good call! Yeah, we need to disallow cross-db references event referencing by id. Fixed that and added test case for it.
Also add cases for renaming a function, altering function's parent schema and dropping a function
</blockquote></details>

Nice fix! See also https://github.com/cockroachdb/cockroach/issues/96674 - we have similar problems with UDTs and tables referenced by OID.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Nice! I just left a couple questions and suggestions.

> Reviewed 16 of 27 files at r1, 9 of 9 files at r2, 1 of 2 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-:-NOAizwj3n86nC3MCMqn:bsczq73)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @DrewKimball, and @shermanCRL)

___
*[`pkg/sql/sem/tree/function_name.go` line 162 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NOAhcYx95OlT1aYVeFH:-NOAhcYx95OlT1aYVeFI:b-ho05b2) ([raw file](https://github.com/cockroachdb/cockroach/blob/92a9df674991e806806cb153b775952ec4e74aef/pkg/sql/sem/tree/function_name.go#L162)):*
> ```Go
> func (*OIDFunctionReference) functionReference()       {}
> 
> type OIDFunctionReference struct {
> ```

nit: Maybe call this just `FunctionOID`.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NNc-MxN4aSKDS8itC2G:-NOAftYM-JzMyXMotIR0:bv3m3oy) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8525d699015f276792ee3068373ae9d86d91a37/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L41)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Nice fix! See also https://github.com/cockroachdb/cockroach/issues/96674 - we have similar problems with UDTs and tables referenced by OID.
</blockquote></details>

Why do we need to disallow cross-database references with this syntax? Are OIDs of functions only unique within a database? That might cause problems when joining two tables from different databases that have computed columns/partial index predictes/etc. that reference UDFs.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 2 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NOAiJ-b82F7y3zmBVKa:-NOAiJ-b82F7y3zmBVKb:b3v052z) ([raw file](https://github.com/cockroachdb/cockroach/blob/92a9df674991e806806cb153b775952ec4e74aef/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L2)):*
> ```
> query T
> SELECT [FUNCTION 1074]('hello,world', ',')
> ```

Can you add a comment explaining what function this is? Is it a builtin?
___
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NOAile1Bw5PdFHDDhlP:-NOAile1Bw5PdFHDDhlQ:b-56lynh) ([raw file](https://github.com/cockroachdb/cockroach/blob/92a9df674991e806806cb153b775952ec4e74aef/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L106)):*
> ```
> 
> statement error pq: cross database function references are not supported
> SELECT [FUNCTION $fn_oid]('hello world')
> ```

Can you add a test with a function OID reference within a UDF? That should error for now.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NNc-MxN4aSKDS8itC2G:-NOAtvgt7gkwDUlXxvak:bhs1gnd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8525d699015f276792ee3068373ae9d86d91a37/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L41)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Why do we need to disallow cross-database references with this syntax? Are OIDs of functions only unique within a database? That might cause problems when joining two tables from different databases that have computed columns/partial index predictes/etc. that reference UDFs.
</blockquote></details>

Thanks for pointing this out, I thought cross-db references are not encouraged in CRDB for some reason. I'll relax this constrain for OID reference since there is a need.


<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NNc-MxN4aSKDS8itC2G:-NOB-Xr6DlodPtQgc6rr:bt01dxn) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8525d699015f276792ee3068373ae9d86d91a37/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L41)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Thanks for pointing this out, I thought cross-db references are not encouraged in CRDB for some reason. I'll relax this constrain for OID reference since there is a need.
</blockquote></details>

I think you're correct about them being discouraged - it's not supported by PG by design, AFAIK. I don't know why we support it - but we do need to make sure that we either produce expected results with cross-db references or have nice errors messages for behaviors we don't support. 


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/sem/tree/function_name.go` line 162 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NOAhcYx95OlT1aYVeFH:-NOBAwxC3S39Qo-QFice:bpmv7dr) ([raw file](https://github.com/cockroachdb/cockroach/blob/92a9df674991e806806cb153b775952ec4e74aef/pkg/sql/sem/tree/function_name.go#L162)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: Maybe call this just `FunctionOID`.
</blockquote></details>

done, good idea


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 2 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NOAiJ-b82F7y3zmBVKa:-NOBB0G52vhLCDypFJQJ:b7ozmj8) ([raw file](https://github.com/cockroachdb/cockroach/blob/92a9df674991e806806cb153b775952ec4e74aef/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L2)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can you add a comment explaining what function this is? Is it a builtin?
</blockquote></details>

yes, it's a builtin, comment added.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NOAile1Bw5PdFHDDhlP:-NOBBBlVEe0id0sQyYDW:b-tnpkcg) ([raw file](https://github.com/cockroachdb/cockroach/blob/92a9df674991e806806cb153b775952ec4e74aef/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L106)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Can you add a test with a function OID reference within a UDF? That should error for now.
</blockquote></details>

thanks for pointing this out, actually fixed a bug where resolver can be nil.


<!-- Sent from Reviewable.io -->

chengxiong-ruan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96442)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball, @mgartner, and @shermanCRL)

___
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NNc-MxN4aSKDS8itC2G:-NOBB8Vn1m2PBA_uP3Gd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8525d699015f276792ee3068373ae9d86d91a37/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L41)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I think you're correct about them being discouraged - it's not supported by PG by design, AFAIK. I don't know why we support it - but we do need to make sure that we either produce expected results with cross-db references or have nice errors messages for behaviors we don't support. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I don't think the behavior of cross-database references has been solved in this PR - that's fine, but let's create an issue to track that work at the least.

> Reviewed 11 of 17 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-:-NOGI94zDlbEHLK3JncW:biaade3)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan and @DrewKimball)

___
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NNc-MxN4aSKDS8itC2G:-NOGHtqI4j8ByW5wlB9B:btebso9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8525d699015f276792ee3068373ae9d86d91a37/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L41)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I'm confused - I don't see tests for these. Did you mean to add them?
___
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 100 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NOGGbC03fy8VuaLaaRS:-NOGGbC1EbetPiqSL-mv:b-v5gnur) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d2977b95c39e52e1780071b6400ded3e586873/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L100)):*
> ```
> SELECT oid FROM pg_catalog.pg_proc WHERE proname = 'f_in_udf'
> 
> statement error pq: function 100110 not found: function undefined
> ```

You need a regex like `\d+` instead of `100110` to prevent this test from flaking, right? Maybe a leave a TODO above this test here too - since we want to allow this in the future, right?
___
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 104 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NOGGpGdBt0yrzzP7SQ0:-NOGGpGdBt0yrzzP7SQ1:bhwjqdz) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2d2977b95c39e52e1780071b6400ded3e586873/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L104)):*
> ```
> 
> statement ok
> CREATE FUNCTION f_using_udf() RETURNS INT LANGUAGE SQL AS $$ SELECT [FUNCTION 814]('abc') $$;
> ```

nit: add a comment explaining that this is allowed because it is a built-in.


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
*[`pkg/sql/logictest/testdata/logic_test/udf_oid_ref` line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/96442#-NNc-MxN4aSKDS8itC2G:-NOGYKOs8pRDTAGoLDEy:b-blzzr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8525d699015f276792ee3068373ae9d86d91a37/pkg/sql/logictest/testdata/logic_test/udf_oid_ref#L41)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm confused - I don't see tests for these. Did you mean to add them?
</blockquote></details>

oh weird, I thought I had a case to test to make sure that's allowed, I'll add one


<!-- Sent from Reviewable.io -->

chengxiong-ruan commented:
TFTRs!
bors r+

chengxiong-ruan commented:
bors r+
